examples/cim/ac/components/experience_shaper.py

```diff
@@ -0,0 +1,48 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from collections import defaultdict
+
+import numpy as np
+
+from maro.rl import ExperienceShaper
+
+
+class TruncatedExperienceShaper(ExperienceShaper):
+    def __init__(self, *, time_window: int, time_decay_factor: float, fulfillment_factor: float,
+                 shortage_factor: float):
+        super().__init__(reward_func=None)
+        self._time_window = time_window
+        self._time_decay_factor = time_decay_factor
+        self._fulfillment_factor = fulfillment_factor
+        self._shortage_factor = shortage_factor
+
+    def __call__(self, trajectory, snapshot_list):
+        agent_ids = np.asarray(trajectory.get_by_key("agent_id"))
+        states = np.asarray(trajectory.get_by_key("state"))
+        actions = np.asarray(trajectory.get_by_key("action"))
+        rewards = np.fromiter(
+            map(self._compute_reward, trajectory.get_by_key("event"), [snapshot_list] * len(trajectory)),
+            dtype=np.float32
+        )
+        return {agent_id: {"states": states[agent_ids == agent_id],
+                           "actions": actions[agent_ids == agent_id],
+                           "rewards": rewards[agent_ids == agent_id],
+                           }
```

Comment:
add next_states

Reply:
The lambda return would require a state series. I think we would be better off passing a state sequence to the algorithm.